packages/extension-url-mapping/src/setup-app.ts

```diff
@@ -0,0 +1,188 @@
+        if (!locals.originalUrl) {
+            return Promise.resolve([])
+        }
+
+        // Example of API call to get url mapping
+        const config = this.getConfig()
+        const shopperSeo = await getShopperSeoClient(locals, config)
+        const urlMapping = await shopperSeo.getUrlMapping({
+            parameters: {urlSegment: locals.originalUrl}
+        })
+        console.log('--- url mapping', urlMapping)
+
+        const requestURL = new URL(locals.originalUrl, getAppOrigin(locals))
+
+        // Returns a partial route
+        return Promise.resolve([
+            {
+                path: requestURL.pathname,
+                componentName: 'Foo',
+                exact: true
+                // component: SamplePage
+            }
+        ])
+    }
+
+    /**
+     * TODO: update comment to make it clearer that beforeRouteMatch is also called on the client side
+     * This method is used on the server during the rendering pipeline. It's provided a list of all the routes that your application
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({allRoutes, locals}: BeforeRouteMatchParams): RouteProps[] {
+        console.log('--- beforeRouteMatch: initial routes', allRoutes)
+        const index = allRoutes.findIndex(
+            (route) =>
+                // @ts-ignore
+                route.componentName === 'Foo'
+        )
+        if (index === -1) {
+            return allRoutes
+        }
+
+        // Complete the partial route
+        const routes = allRoutes.slice()
+        const [myRoute] = routes.splice(index, 1)
+        myRoute.component = routeComponent(SamplePage, true, locals)
+        // NOTE: to be expected: the Sample page will be rendered on the server side, while a different page is then rendered on the client side.
+        // Jinsu's work on serialization will make sure that the same page will be rendered on both server and client sides.
+
+        const result = [myRoute, ...routes]
+        console.log('--- beforeRouteMatch: resulting routes', result)
+        return result
+    }
+}
+
+export default UrlMapping
+
+const getShopperSeoClient = async (locals: Record<string, any>, config: Config) => {
+    const {
+        commerceAPI,
+        commerceAPIAuth: {propertyNameInLocals: authProperty}
+    } = config
+
+    const appOrigin = getAppOrigin(locals)
+
+    // Saving/reusing the commerce api auth (so all extensions have access to it)
+    locals[authProperty] =
+        locals[authProperty] ??
+        new Auth({
+            ...commerceAPI.parameters,
+            proxy: `${appOrigin}${commerceAPI.proxyPath}`,
+            redirectURI: `${appOrigin}/callback`,
+            logger: console // TODO: proper logger
+        })
```

Comment:
Which package should be the one handling the commerce API stuffs? Initially I thought one of our SDKs (react or extension SDKs). But they don't have a dependency on commerce-sdk-react.

So I thought it's better that url-mapping extension is the one that does it. And the SDK can pass in the `locals` object, which is useful for sharing the auth/access token.